diff --color -rupN base/linux-5.4.109/Makefile linux-5.4.109/Makefile
--- base/linux-5.4.109/Makefile	2021-03-30 05:35:30.000000000 -0700
+++ linux-5.4.109/Makefile	2024-07-04 17:10:21.491275586 -0700
@@ -1036,7 +1036,7 @@ PHONY += prepare0
 export MODORDER := $(extmod-prefix)modules.order
 
 ifeq ($(KBUILD_EXTMOD),)
-core-y		+= kernel/ certs/ mm/ fs/ ipc/ security/ crypto/ block/
+core-y		+= kernel/ certs/ mm/ fs/ ipc/ security/ crypto/ block/ cmpt300/
 
 vmlinux-dirs	:= $(patsubst %/,%,$(filter %/, $(init-y) $(init-m) \
 		     $(core-y) $(core-m) $(drivers-y) $(drivers-m) \
diff --color -rupN base/linux-5.4.109/arch/x86/entry/syscalls/syscall_64.tbl linux-5.4.109/arch/x86/entry/syscalls/syscall_64.tbl
--- base/linux-5.4.109/arch/x86/entry/syscalls/syscall_64.tbl	2021-03-30 05:35:30.000000000 -0700
+++ linux-5.4.109/arch/x86/entry/syscalls/syscall_64.tbl	2024-07-04 21:11:15.506879555 -0700
@@ -357,6 +357,8 @@
 433	common	fspick			__x64_sys_fspick
 434	common	pidfd_open		__x64_sys_pidfd_open
 435	common	clone3			__x64_sys_clone3/ptregs
+437	common  array_stats		__x64_sys_array_stats
+438	common  process_ancestors	__x64_sys_process_ancestors
 
 #
 # x32-specific system call numbers start at 512 to avoid cache impact
diff --color -rupN base/linux-5.4.109/cmpt300/Makefile linux-5.4.109/cmpt300/Makefile
--- base/linux-5.4.109/cmpt300/Makefile	1969-12-31 16:00:00.000000000 -0800
+++ linux-5.4.109/cmpt300/Makefile	2024-07-04 21:56:57.426046458 -0700
@@ -0,0 +1 @@
+obj-y := array_stats.o process_ancestors.o
diff --color -rupN base/linux-5.4.109/cmpt300/array_stats.c linux-5.4.109/cmpt300/array_stats.c
--- base/linux-5.4.109/cmpt300/array_stats.c	1969-12-31 16:00:00.000000000 -0800
+++ linux-5.4.109/cmpt300/array_stats.c	2024-07-05 00:30:50.671125022 -0700
@@ -0,0 +1,72 @@
+#include <linux/kernel.h>
+#include <linux/syscalls.h>
+#include <linux/uaccess.h>
+#include "array_stats.h"
+
+SYSCALL_DEFINE3(array_stats, struct array_stats *, user_stats, long *, user_data, long, size){
+	// initialize variables
+	int i = 0;
+	long temp = 0;
+	long min = 0;
+	long max = 0;
+	long sum = 0;
+	struct array_stats stats;
+
+	// check if size is 0
+	if (size <= 0) {
+		printk("Error: size of array is <= 0\n");
+		return -EINVAL;
+	}
+
+	// check if there is any error accessing stats or data pointers
+	else if (!access_ok(user_stats, sizeof(struct array_stats)) || !access_ok(user_data, sizeof(long) * size)) {
+		printk("Error: problem with accessing user array or user data\n");
+		return -EFAULT;
+	}
+
+	// find stat values
+	else {
+		for (i = 0; i < size; i++) {
+			if (copy_from_user(&temp, user_data + i, sizeof(long))) {
+			printk("Error: copy_from_user failed copying from array\n");
+			return -EFAULT;
+			}
+			if (i == 0) {
+				min = temp;
+				max = temp;
+				sum = temp;
+			}
+			// replace min or max with temp if conditions meet
+			else {
+				if (temp <= min) {
+					min = temp;
+				}
+
+				if (temp >= max) {
+					max = temp;
+				}
+				// add to sum
+				sum += temp;
+			}
+		}
+
+		// add values into struct
+		stats.min = min;
+		stats.max = max;
+		stats.sum = sum;
+
+		// move values into user struct
+		if (copy_to_user(user_stats, &stats, sizeof(struct array_stats))) {
+			printk("Error: copy_to_user failed copying stats into user struct\n");
+			return -EFAULT;
+		}
+
+		// print data values in kernal space
+		printk("Min: %ld\n", stats.min);
+		printk("Max: %ld\n", stats.max);
+		printk("Sum: %ld\n", stats.sum);
+
+		return 0;
+	}
+
+}
diff --color -rupN base/linux-5.4.109/cmpt300/array_stats.h linux-5.4.109/cmpt300/array_stats.h
--- base/linux-5.4.109/cmpt300/array_stats.h	1969-12-31 16:00:00.000000000 -0800
+++ linux-5.4.109/cmpt300/array_stats.h	2024-07-04 17:45:42.550630124 -0700
@@ -0,0 +1,14 @@
+// Define the array_stats struct for the array_stats syscall
+#ifndef _ARRAY_STATS_H_
+#define _ARRAY_STATS_H_
+
+struct array_stats {
+	long min;
+	long max;
+	long sum;
+};
+
+asmlinkage long sys_array_stats(struct array_stats *stats, long *data, long size);
+
+#endif
+
diff --color -rupN base/linux-5.4.109/cmpt300/process_ancestors.c linux-5.4.109/cmpt300/process_ancestors.c
--- base/linux-5.4.109/cmpt300/process_ancestors.c	1969-12-31 16:00:00.000000000 -0800
+++ linux-5.4.109/cmpt300/process_ancestors.c	2024-07-05 00:42:23.512161197 -0700
@@ -0,0 +1,68 @@
+#include <linux/kernel.h>
+#include <linux/syscalls.h>
+#include <linux/sched.h>
+#include <linux/uaccess.h>
+#include "process_ancestors.h"
+
+SYSCALL_DEFINE3(process_ancestors, struct process_info *, info_array, long, size, long *, num_filled) {
+    long process_count = 0;
+    struct task_struct *p;
+    struct process_info proc_info; /*struct for storing process info in kernel space*/
+    
+    // handle size <= 0 
+    if (size <= 0) {
+		printk("Error: size of array is <= 0\n");
+		return -EINVAL;
+	}
+
+    // handle problems with accessing info_array or num_filled
+    else if (!access_ok(info_array, sizeof(struct process_info) * size) || !access_ok(num_filled, sizeof(long))) {
+		printk("Error: problem with accessing info_array or num_filled\n");
+		return -EFAULT;
+	}
+    
+    else {
+        p = current; // point to current process
+
+        /*loop unless array full or searched fully in process tree*/
+        while (process_count < size && p->parent != p) {
+            struct list_head *node;
+
+            proc_info.pid = p->pid;
+            strncpy(proc_info.name, p->comm, ANCESTOR_NAME_LEN - 1);
+            proc_info.name[ANCESTOR_NAME_LEN - 1] = '\0';
+            proc_info.state = p->state;
+            proc_info.uid = p->real_cred->uid.val;
+            proc_info.nvcsw = p->nvcsw;
+            proc_info.nivcsw = p->nivcsw;
+
+            // source : https://stackoverflow.com/questions/36536274/calculating-the-number-of-children-and-sibiling-of-a-process-in-the-kernel-mode (Author:falhumai)
+            // count children
+            proc_info.num_children = 0;
+            list_for_each(node, &(p->children)) {
+                proc_info.num_children++;
+            }
+            // count siblings
+            proc_info.num_siblings = 0;
+            list_for_each(node, &(p->sibling)) {
+                proc_info.num_siblings++;
+            }
+
+            // move to user space
+            if (copy_to_user(&info_array[process_count], &proc_info, sizeof(struct process_info))) {
+                printk("Error: Fail to copy to user space array\n");
+                return -EFAULT;
+            }
+            process_count++;
+            p = p->parent;
+        }
+
+        // move num_filled to user space
+        if (copy_to_user(num_filled, &process_count, sizeof(long))) {
+            printk("Error: Fail to copy count into num_filled\n");
+            return -EFAULT;
+        }
+
+        return 0;
+    }
+}
\ No newline at end of file
diff --color -rupN base/linux-5.4.109/cmpt300/process_ancestors.h linux-5.4.109/cmpt300/process_ancestors.h
--- base/linux-5.4.109/cmpt300/process_ancestors.h	1969-12-31 16:00:00.000000000 -0800
+++ linux-5.4.109/cmpt300/process_ancestors.h	2024-07-04 22:44:04.237600914 -0700
@@ -0,0 +1,19 @@
+// Structure to hold values returned by process_ancestors sys-call
+#ifndef _PROCESS_ANCESTORS_H_
+#define _PROCESS_ANCESTORS_H_
+
+#define ANCESTOR_NAME_LEN 16
+struct process_info {
+    long pid;
+    char name[ANCESTOR_NAME_LEN];
+    long state;
+    long uid;
+    long nvcsw;
+    long nivcsw;
+    long num_children;
+    long num_siblings;
+};
+
+asmlinkage long sys_process_ancestors(struct process_info *info_array, long size, long *num_filled);
+
+#endif
\ No newline at end of file
Binary files base/linux-5.4.109/tools/objtool/objtool and linux-5.4.109/tools/objtool/objtool differ
